### PR TITLE
Added Toast component

### PR DESCRIPTION
**Pull request by benja**
image

This PR introduces the Toast component to React.

It supports manual broadcasting by using the `<Toast>` component, while still allowing for you to send off toast alerts programmatically using the `useToast()` hook as seen in the example below.

### Manual toasting
```jsx
<Toast type="success" text="message" />
```

### Programmatic toasting

```jsx
const { toast } = useToast();

toast('message', {
    type: 'success',
    // ...other options
})
```

Currently available options:

- type (string) — type of alert: `success | error | warning | info`
- canClose (bool) — Can toast be dismissed?
- onClose (func) — Dismiss handler
- duration (time in ms) — Duration of toast in milliseconds. Set to 0 to disable auto-removal

[Read docs](https://opensource.finn.no/fabric-react/#/toast)

Example video demonstrating programmatic toasts.

https://user-images.githubusercontent.com/16625915/126485202-82aee617-ab10-4cef-b746-83968ced481e.mov

### CODE REVIEW

**Comment by maughan** [comment]
lgtm! 👀

File: packages/toast/src/toast.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        <div
+            className={classNames(c.toastWrapper, {
+                expandIn: isProgrammatic && !disappeared,
+                expandOut: isProgrammatic && disappeared,
+            })}
+            role="status"
+            aria-live="polite"
+        >
+            <div
+                id={`toast-${props.id}-wrapper`}
```
Comment (maughan) [inline packages/toast/src/toast.tsx:51]: Hey Benja, looks good!

I did have a small concern here as you typed id as an optional param, making the above potentially undefined. 

`}: ToastProps & { id?: string; programmatic?: boolean }) {`

**Review by maughan — CHANGES_REQUESTED** [review]

File: packages/toast/src/toast.tsx (hunk 1 above)
Comment (fifty) [reply packages/toast/src/toast.tsx:51]: `
return {
        toast: (message: string, options: Omit<ToastProps, 'text'>) => {
            const id =
                Date.now().toString(36) +
                Math.random().toString(36).slice(2, 5);
`

This is a good point. The id regardless of whether it is passed will fallback to a custom identifier therefore making it impossible outside of the scope of the context to delete the toast (could be used in some use cases)

**Review by digitalsadhu — COMMENTED** [review]
This looks really nice. I can see that individual toasts can be added to specific areas of a page by using the component directly and that the global toast can be manipulated programmatically etc. Does that continue to work if say 3 podlets on the page all programmatically manipulate the toast separately?

**Comment by benja** [comment]
> This looks really nice. I can see that individual toasts can be added to specific areas of a page by using the component directly and that the global toast can be manipulated programmatically etc. Does that continue to work if say 3 podlets on the page all programmatically manipulate the toast separately?

That is a great point. Currently, if there are multiple podlets running on the same page they will all use their own toast context. This means that any application running outside of the scope of the React application won't be able to manipulate its toasts. If two podlets fire off a toast at the same time they would just overlap, which is not the expected behavior.

The problem lies in having a global context cross-application. I believe one way to achieve this would be to store toasts on `globalThis` / `window`, which means any application running on the same page will have access to the same toast context. We could then provide a similar API to push, edit and remove toasts from this context.

We also spoke about moving the Toast component to a custom element, which would solve the duplicate implementation, but we'd still have the same problem of global toast context, unless we use a global context that's accessible cross-framework, whether that be through messages or using `globalThis` / `window`

File: packages/toast/src/toast.tsx (hunk 1 above)
Comment (benja) [reply packages/toast/src/toast.tsx]: > `return { toast: (message: string, options: Omit<ToastProps, 'text'>) => { const id = Date.now().toString(36) + Math.random().toString(36).slice(2, 5);`
> 
> This is a good point. The id regardless of whether it is passed will fallback to a custom identifier therefore making it impossible outside of the scope of the context to delete the toast (could be used in some use cases)

The identifier is only generated for programmatic toasts and won't be there if you're using `<Toast>` individually. Although, this does raise another question about a global shared context between multiple applications on the same page. We use [podlets](https://podium-lib.io/docs/api/podlet) which is its own running application and can be just a fragment of a full page.

Currently, each podlet running React and using toasts would create its own instance of the toast context, and if two or more podlets fire off a toast at the same time they would all just overlap. Whether we solve this through sockets, `window` or some sort of messaging is a question to be answered.

Comment (benja) [reply packages/toast/src/toast.tsx]: Great point, @fifty. I've added the option to pass a custom identifier when programmatically creating a toast.

**Comment by benja** [comment]
We've gone ahead and implemented the Toast in [Fabric Elements](https://opensource.finn.no/fabric-elements/) using custom elements. This implementation uses window as the global state and all applications on the same page, no matter which framework is being used, will be able to utilise the same API to send off and manipulate toasts.

See module: https://www.npmjs.com/package/@finn-no/toast.

**Pull request closed without merging**